### PR TITLE
Sync @JsonKey options with freezed decorators

### PR DESCRIPTION
**Pull request by mseijas**
Hi @rrousselGit , 

Our team has run into some issues where the freezed annotations that we define sometimes either clash with the default `@JsonKey` settings we get, or we're forced to duplicate some semantics around how we want our models to operate.

For example:
```dart
class User with _$User {
 	factory User({
		@required @JsonKey(required: true, disallowNullValues: true) name,
		@nullable @JsonKey(disallowNullValues: false) phoneNumber,
	}) = _User;
}
```

It would be great if we could look at the `@required` and `@nullable` annotations and make sure that the options set on `@JsonKey` also reflect those settings.

I took a stab at this, would love to know your thoughts!

### CODE REVIEW

File: packages/freezed/lib/src/templates/concrete_template.dart [hunk 1]
```diff
@@ -370,25 +369,90 @@ class Property {
...
+    }
+
+    return '${decorators.join()} $jsonKeyOptions final $type $name;';
   }
 
   Getter get getter => Getter(
       name: name, type: type, decorators: decorators, nullable: nullable);
 }
 
+class JsonKeyOptions {
```
Comment (rrousselGit) [inline packages/freezed/lib/src/templates/concrete_template.dart:414]: This should be done by the Property class instead

File: packages/freezed/lib/src/templates/concrete_template.dart [hunk 2]
```diff
@@ -370,25 +369,90 @@ class Property {
...
+  final Map<String, String> values;
+
+  JsonKeyOptions([this.values = const {}]);
+
+  static JsonKeyOptions fromDecorators(List<String> decorators) {
+    if (decorators == null || decorators.isEmpty) return null;
+
+    String jsonKeyOptions;
+    for (final d in decorators) {
+      if (!d.startsWith('@JsonKey')) continue;
```
Comment (rrousselGit) [inline packages/freezed/lib/src/templates/concrete_template.dart:424]: There are extensions already for such thing. Search for the extensions on the project, you should find them

This shouldn't be done with a `.split`

**Comment by rrousselGit** [comment]
Could you add tests for this PR?

File: packages/freezed/lib/src/templates/concrete_template.dart (hunk 1 above)
Comment (mseijas) [reply packages/freezed/lib/src/templates/concrete_template.dart:414]: 👍

**Comment by mseijas** [comment]
> Could you add tests for this PR? And fix the CI too

Yeah for sure - just wanted to make sure this seemed like it was going in the right direction to you first. Will add!

**Comment by mseijas** [comment]
@rrousselGit Updated the implementation based on your suggestions, and also added tests. Let me know what you think.

File: packages/freezed/lib/src/templates/concrete_template.dart [hunk 3]
```diff
@@ -370,41 +367,123 @@ class Property {
       );
     }
 
+    final jsonKeyValues = element.jsonKeyValues;
+    if (defaultValue != null &&
+        jsonKeyValues != null &&
+        jsonKeyValues.containsKey('defaultValue')) {
+      throw InvalidGenerationSourceError(
+        '@Default cannot be used in conjunction with @JsonKey(defaultValue:)',
```
Comment (rrousselGit) [inline packages/freezed/lib/src/templates/concrete_template.dart:375]: Why is this needed?

File: packages/freezed/lib/src/templates/concrete_template.dart [hunk 4]
```diff
@@ -370,41 +367,123 @@ class Property {
...
+        todo: 'Remove the defaultValue parameter from @JsonKey and only'
+            'use the @Default decorator instead',
+        element: element,
+      );
+    }
+
+    if (element.hasRequired &&
+        jsonKeyValues != null &&
+        jsonKeyValues.containsKey('required')) {
+      throw InvalidGenerationSourceError(
```
Comment (rrousselGit) [inline packages/freezed/lib/src/templates/concrete_template.dart:385]: Same here, why is this needed?

File: packages/freezed/test/json_test.dart [hunk 5]
```diff
@@ -28,10 +29,38 @@ Future<void> main() async {
       DefaultValue(42),
     );
   });
-  test('@Default does not imply a @JsonKey if one is already specified', () {
+  test('@required implies a @JsonKey', () {
```
Comment (rrousselGit) [inline packages/freezed/test/json_test.dart:32]: What is the purpose of this behavior?

Before, it threw a null exception anyway, so I don't see what this adds

File: packages/freezed/test/json_test.dart [hunk 6]
```diff
@@ -28,10 +29,38 @@ Future<void> main() async {
...
+      () => RequiredValue.fromJson(<String, dynamic>{}),
+      throwsA(isA<CheckedFromJsonException>()),
+    );
+
+    expect(
+      () => RequiredValue.fromJson(<String, dynamic>{'value': null}),
+      throwsA(isA<CheckedFromJsonException>()),
+    );
+  });
+  test('@nullable implies a @JsonKey', () {
```
Comment (rrousselGit) [inline packages/freezed/test/json_test.dart:49]: I don't understand what this test does, since this is already the current behavior

File: packages/freezed/lib/src/templates/concrete_template.dart (hunk 3 above)
Comment (mseijas) [reply packages/freezed/lib/src/templates/concrete_template.dart:375]: So I added this for safety, to prevent someone writing something like this:
```dart
factory User({@Default(42) @JsonKey(defaultValue: 30) int age})
```
At the call site, it is unclear what behavior will result from this.

File: packages/freezed/lib/src/templates/concrete_template.dart (hunk 4 above)
Comment (mseijas) [reply packages/freezed/lib/src/templates/concrete_template.dart:385]: This is because you could technically write something like this which would also be unclear in terms of what the outcome will be
```dart
factory User({@required @JsonKey(required: false) int age})
```

File: packages/freezed/test/json_test.dart (hunk 5 above)
Comment (mseijas) [reply packages/freezed/test/json_test.dart:32]: So before it threw an `AssertionError` -- which is fine in debug builds, but would not get thrown in release builds. This would mean that if deserialization failed in production you would have no way of knowing and the app could end up in a bad state.

What this new behavior does is produce a `CheckedFromJsonException`-- meaning that, since it's not an assertion error, it would still be produced in release builds.

File: packages/freezed/test/json_test.dart (hunk 6 above)
Comment (mseijas) [reply packages/freezed/test/json_test.dart:49]: Fair, maybe this is not the way to test this. I just wanted to make sure that if we had the `@nullable` decorator that we would add the corresponding decorator: `@JsonKey(required: false, disallowNullValue: false)`

Do you have any suggestions on how I can test this?

Comment (mseijas) [reply packages/freezed/test/json_test.dart:49]: For added context, the reason why I think this is important is because someone could write:
```dart
factory User({@nullable @JsonKey(disallowNullValue: true) int age})
```
Resulting again in an indeterministic behavior

**Comment by rrousselGit** [comment]
To be honest, I still don't quite understand what you are trying to do.
Could you explain in further detail the problem?

Maybe what you want is instead to change the default behavior of `@JsonKey` to match `@required` and `@nullable`?

File: packages/freezed/test/json_test.dart (hunk 6 above)
Comment (rrousselGit) [reply packages/freezed/test/json_test.dart:49]: The behavior is fixed: `@JsonKey` inside the constructor explicitly takes over every other configuration.

File: packages/freezed/test/json_test.dart (hunk 5 above)
Comment (rrousselGit) [reply packages/freezed/test/json_test.dart:32]: That's a fair point, but it's unrelated to deserialisation.
If we want to throw errors on release, this should happen for all usages, not just `.fromJson`

It could be a different mode where instead of:

```dart
MyClass(int foo): assert(foo != null);
```

we get:

```dart
MyClass(int foo) {
  if (foo == null) throw StateError('foo cannot be null');
}
```

Comment (mseijas) [reply packages/freezed/test/json_test.dart:32]: Yes! I considered that too. It just felt annoying that then we couldn't have `const` constructors anymore 😞 But I hear your point! In the end, and to keep things simpler, I thought that throwing errors in deserialization was more important because those models get initialized from data that comes from the server and we don't control. So the chances of something going wrong there seemed higher than the chances of manually building a model passing in invalid values.

**Comment by mseijas** [comment]
> Maybe what you want is instead to change the default behavior of @JsonKey to match @required and @nullable?

Yes, exactly! That's what I want to do. 

Basically, I just want to depend on using the `@freezed` annotations and not have to duplicate those definitions with `@JsonKey` annotations as well.

So, if I use a `@required` annotation, I don't want to have to manually also write `@JsonKey(required: true, disallowNullValue: true)`

The same thing with the `@nullable` annotation, I don't want to have to manually also add `@JsonKey(required: false, disallowNullValue: false)`

Additionally, I thought it would also be nice that I didn't have `@freezed` and `@JsonKey` annotations that contradict themselves, like:
`@Default(42) @JsonKey(defaultValue: 30)`

Does that make sense? I'm happy to explore other ways of achieving this, but that was what I was aiming for.

File: packages/freezed/test/json_test.dart (hunk 5 above)
Comment (rrousselGit) [reply packages/freezed/test/json_test.dart:32]: If the exception happens during deserialisation, const constructors don't matter

The benefit is, this continues to work on copyWith too

File: packages/freezed/test/json_test.dart (hunk 6 above)
Comment (mseijas) [reply packages/freezed/test/json_test.dart:49]: But don't you think that this is problematic when working in a big team, and looking at someone else's code that you're not familiar with?

Although the behavior might be fixed based on what happens at run time, when reading this code it is not immediately clear which one of those decorators should superseed the other. I think it might be nice to prevent users from defining decorators that could in theory contradict each other, which means that one of them would actually end up being ignored.

Comment (rrousselGit) [reply packages/freezed/test/json_test.dart:49]: To me, it doesn't matter, as people shouldn't need to specific `@JsonKey` to begin with.

The support for `@JsonKey` in parameters is here only for people want to override the default behavior. So in that situation, it is clear that it takes over everything else.

What you want is not to check for inconsistencies. You want to change the default behavior.

File: packages/freezed/test/json_test.dart (hunk 5 above)
Comment (mseijas) [reply packages/freezed/test/json_test.dart:32]: Ok, I'm happy to explore that approach instead if you think that's better!

File: packages/freezed/test/json_test.dart (hunk 6 above)
Comment (mseijas) [reply packages/freezed/test/json_test.dart:49]: But there's also the case where you might want to specify a `@JsonKey` just so that you get a `CheckedFromJsonException` instead of a runtime assertion error. (Although if we modify our constructors the way you suggested above then this wouldn't matter anymore)

I do think it might still be nice to throw a `InvalidGenerationSourceError` during codegen when you have defined decorators that contradict each other as a sanity check? What do you think of that?

**Comment by rrousselGit** [comment]
We could change the default `@JsonKey` such that if the property is nullable,  it is generated as `@JsonKey(nullable: true)`

I'm not quite sure why we would want `required` or `disallowNullValue` instead.

File: packages/freezed/test/json_test.dart (hunk 6 above)
Comment (rrousselGit) [reply packages/freezed/test/json_test.dart:49]: > I do think it might still be nice to throw a InvalidGenerationSourceError during codegen when you have defined decorators that contradict each other as a sanity check? What do you think of that?

I would agree if people were expected to write `@JsonKey(something: true)` themselves. But they are not
As I mentioned, `@JsonKey` on parameters is only here for edge-cases to let people have fine grained control over the annotation.

Comment (mseijas) [reply packages/freezed/test/json_test.dart:49]: > As I mentioned, @JsonKey on parameters is only here for edge-cases to let people have fine grained control over the annotation.

Of course, yeah. I just meant to prevent errors from slipping through when dealing with these edge-cases.

**Comment by mseijas** [comment]
> We could change the default `@JsonKey` such that if the property is nullable, it is generated as `@JsonKey(nullable: true)`
> 
> I'm not quite sure why we would want `required` or `disallowNullValue` instead.

So based on the json_serialization I think the `@JsonKey` properties should be set like this:

With freezed `@required` annotation, it should have the following `@JsonKey` settings:
`required: true` -> When true, generated code for fromJson will verify that the source JSON map contains the associated key.
`disallowNullValue: true` -> If true, generated code will throw a DisallowedNullValueException if the corresponding key exists, but the value is null.

With freezed `@nullable` annotation, it should have the following `@JsonKey` settings:
`required: false` -> When false, won't verify that source JSON map contains the associated key.

`nullable: true` works strangely actually, check out the documentation for it:
When true, generated code for fromJson will verify that the source JSON map contains the associated key.

If the key does not exist, a MissingRequiredKeysException exception is thrown.

**Comment by mseijas** [comment]
That said, would you prefer for me to explore the approach of updating our constructors to throw any necessary exceptions instead of this other technique of creating/updating the `@JsonKey` settings?

**Comment by rrousselGit** [comment]
> That said, would you prefer for me to explore the approach of updating our constructors to throw any necessary exceptions instead of this other technique of creating/updating the `@JsonKey` settings?

That would be preferable yes.

Ideally, the default behavior should stay as-is. And then, we'd have a setting on the `build.yaml` file & on `@Freezed()` to change the behavior.

**Comment by mseijas** [comment]
> > That said, would you prefer for me to explore the approach of updating our constructors to throw any necessary exceptions instead of this other technique of creating/updating the `@JsonKey` settings?
> 
> That would be preferable yes.
> 
> Ideally, the default behavior should stay as-is. And then, we'd have a setting on the `build.yaml` file & on `@Freezed()` to change the behavior.

Cool! I'll explore that option 👍

**Comment by mseijas** [comment]
Something to consider for later, I think we might still want to at least print a warning when doing something like this:

```dart
factory User({@Default(42) @JsonKey(defaultValue: 30) int age})
```

Since doing that really doesn't make sense. Right?

**Comment by rrousselGit** [comment]
Warnings would be fine. But it's not the role of the code-generator.

We need a custom lint rule for this instead. So that they can be ignored or
disabled.

This is feasible, but is a different project

**Comment by mseijas** [comment]
> We need a custom lint rule for this instead. So that they can be ignored or
disabled.

Oh I like that!

**Comment by rrousselGit** [comment]
After I publish a project I'm working on, I'll try and make a custom linter
for all my different projects (flutter_hooks, freezed, provider, ...)

So this may come in due time. But it is a different issue

Le jeu. 14 mai 2020 à 17:47, Matias Seijas <notifications@github.com> a
écrit :

**Comment by rrousselGit** [comment]
Do you still plan on working on this, or should we close this PR?

**Comment by mseijas** [comment]
Yes! I haven't gotten around to it yet, but I will soon. Happy to open a different PR if you prefer though.
